### PR TITLE
Implement agency dashboard component and handle redirections

### PR DESCRIPTION
**Pull request by yashwin**
#### Changes proposed in this Pull Request

This PR adds a header, sidebar, and component(dashboard) for the upcoming Jetpack Agency Dashboard and the redirection logic for the same.

The designs are currently in flux, so this PR only displays a text saying `Site overview` to confirm that everything is set up correctly and will be used as the foundation for the upcoming PRs as a part of the Jetpack Agency Dashboard.

#### Testing instructions

**Prerequisites**

Since this change is made specifically for agencies, you will have to set yourself(partner) as an agency - 2c49b-pb

**Instructions**

1. Run `git checkout add/implement-agency-dashboard-component` and `yarn start-jetpack-cloud`
2. Apply D79392-code (also take a look at its testing instructions)
3. Visit http://jetpack.cloud.localhost:3000, and confirm that you are being redirected to  http://jetpack.cloud.localhost:3000/dashboard and `Site overview` is displayed.
5. Check if the sidebar and header are displayed
6. In `config/jetpack-cloud-development.json` set the `jetpack/agency-dashboard` flag to `false`, rebuild and reload the dashboard page, and verify that you are redirected to http://jetpack.cloud.localhost:3000/ (which will further redirect to `http://jetpack.cloud.localhost:3000/landing` in most cases).

`Note` We have added a header component that is the same as the header component which already exists in the partner portal, this was added since the header in the dashboard is subjected to change. 

Screenshot 2022-04-19 at 9 57 36 AM

Related to 1202076982646589-as-1202117620005342 & 1202076982646589-as-1202122273254220

### CODE REVIEW

File: client/jetpack-cloud/index.js [hunk 1]
```diff
@@ -26,15 +27,22 @@ const clearPageTitle = ( context, next ) => {
...
+const redirectToPrimarySiteLanding = ( context, next ) => {
 	debug( 'controller: redirectToPrimarySiteLanding', context );
 	const state = context.store.getState();
 	const currentUser = getCurrentUser( state );
 	const isPrimarySiteJetpackSite = getPrimarySiteIsJetpack( state );
 
-	isPrimarySiteJetpackSite
-		? page( `/landing/${ currentUser.primarySiteSlug }` )
-		: page( `/landing` );
+	const isAgency = config.isEnabled( 'jetpack/agency-dashboard' );
```
Comment (rcanepa) [inline client/jetpack-cloud/index.js]: The `isAgency` needs to consider also whether the user is an agency (this is something that folks at Avalon might know how to do). Otherwise, if this feature flag is enabled, all users will be redirected to this section regardless of their type. This will create a lot of noise for internal users visiting https://cloud.jetpack.com who are not aware of these changes.

File: client/jetpack-cloud/sections/agency-dashboard/header/index.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+			<div className={ classNames( 'header' ) }>
+				<FormattedHeader
+					className="header__main-title"
+					headerText={ preventWidows( title ) }
+					align="center"
+				/>
+			</div>
+		</>
+	);
+}
```
Comment (rcanepa) [inline client/jetpack-cloud/sections/agency-dashboard/header/index.tsx:28]: I see we already have a component that looks exactly like this (see https://github.com/Automattic/wp-calypso/blob/49b24160513c22585c26e66b39bcbeff74a62fa7/client/jetpack-cloud/sections/partner-portal/header/index.tsx#L10). Can we use the same in both sections? I suspect we would have to move the component up the hierarchy to a place that would make sense.

File: client/jetpack-cloud/sections/agency-dashboard/sidebar/index.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	};
+
+	const onGetHelp = () => {
+		reduxDispatch(
+			recordTracksEvent( 'calypso_jetpack_sidebar_menu_click', {
+				menu_item: 'Jetpack Cloud / Support',
+			} )
+		);
+		window.scrollTo( 0, 0 );
+	};
```
Comment (rcanepa) [inline client/jetpack-cloud/sections/agency-dashboard/sidebar/index.tsx]: I think we can use `onNavigate` instead of this function since they both track the same event and do the same behavior.

File: client/jetpack-cloud/sections/agency-dashboard/sidebar/index.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+				<SidebarFooter>
+					<SidebarMenu>
+						<SidebarItem
+							label={ translate( 'Get help', {
+								comment: 'Jetpack Cloud sidebar navigation item',
+							} ) }
+							link="https://jetpack.com/support"
+							className="sidebar__jetpack-cloud-item-has-border"
+							customIcon={ <JetpackIcons icon="help" /> }
+							onNavigate={ onGetHelp }
```
Comment (rcanepa) [inline client/jetpack-cloud/sections/agency-dashboard/sidebar/index.tsx]: Here we can use `onNavigate( 'Jetpack Cloud / Support' );`.

**Comment by rcanepa** [comment]
I noticed that when one clicks on Switch Site > All My Sites, one is not redirected to the Dashboard, which is part of the expected behavior of these changes.

File: client/my-sites/controller.js [hunk 5]
```diff
@@ -547,21 +547,26 @@ export function navigation( context, next ) {
...
-	if ( context.query.verified === '1' ) {
-		context.store.dispatch(
-			successNotice(
-				i18n.translate(
-					"Email verified! Now that you've confirmed your email address you can publish posts on your blog."
+	const isAgencyDashboardEnabled = config.isEnabled( 'jetpack/agency-dashboard' );
+	if ( isAgencyDashboardEnabled ) {
+		page.redirect( '/dashboard' );
+	} else {
+		if ( context.query.verified === '1' ) {
```
Comment (rcanepa) [inline client/my-sites/controller.js]: What is the purpose of this check?

File: client/jetpack-cloud/index.js [hunk 6]
```diff
@@ -26,15 +27,22 @@ const clearPageTitle = ( context, next ) => {
...
+	const isAgencyDashboardEnabled = config.isEnabled( 'jetpack/agency-dashboard' );
+
+	if ( isAgencyDashboardEnabled ) {
+		page( `/dashboard` );
+	} else {
+		isPrimarySiteJetpackSite
+			? page( `/landing/${ currentUser.primarySiteSlug }` )
+			: page( `/landing` );
+	}
+	next();
```
Comment (rcanepa) [inline client/jetpack-cloud/index.js:41]: What is the purpose of calling `next()` since every code path leads the user to somewhere via a call to `page(...)`?

File: client/my-sites/controller.js [hunk 7]
```diff
@@ -547,21 +547,26 @@ export function navigation( context, next ) {
...
 export function sites( context, next ) {
-	if ( context.query.verified === '1' ) {
-		context.store.dispatch(
-			successNotice(
-				i18n.translate(
-					"Email verified! Now that you've confirmed your email address you can publish posts on your blog."
+	const isAgencyDashboardEnabled = config.isEnabled( 'jetpack/agency-dashboard' );
+	if ( isAgencyDashboardEnabled ) {
+		page.redirect( '/dashboard' );
+	} else {
```
Comment (mattgawarecki) [inline client/my-sites/controller.js]: For safety, I recommend adding a `return` statement anywhere we redirect -- [`page.redirect` is an asynchronous call](https://github.com/visionmedia/page.js/blob/master/page.js#L660), so it's possible other routing code will execute after we decide to redirect. We can also then unwrap the enclosing `else` block immediately after:

```suggestion
	if ( isAgencyDashboardEnabled ) {
		page.redirect( '/dashboard' );
		return;
```

File: client/jetpack-cloud/index.js (hunk 6 above)
Comment (yashwin) [reply client/jetpack-cloud/index.js:41]: Good catch, I'll remove this.

File: client/my-sites/controller.js (hunk 5 above)
Comment (yashwin) [reply client/my-sites/controller.js]: This is an existing check, I'm not entirely sure what it does.

File: client/my-sites/controller.js (hunk 7 above)
Comment (yashwin) [reply client/my-sites/controller.js]: Makes sense, will make the suggested change.

File: client/my-sites/controller.js (hunk 5 above)
Comment (yashwin) [reply client/my-sites/controller.js]: By looking at the code, this check is there to show the success notice when email verification is done. So, I'm gonna move the above check below this block.

**Comment by rcanepa** [comment]
This works much better now. However, I noticed that when I go from a single site section (e.g., Scan) to Switch Site > All My Sites, I see a quick flash of a header that says `You are not authorized to see this page`, which makes me think there is a race condition within the check that decides what to show to the user.

File: client/lib/partner/utils.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+export const getShowAgencyDashboard = ( store ) => {
+	const state = store.getState();
+	let isAgency;
+	if ( config.isEnabled( 'jetpack/agency-dashboard' ) ) {
+		const partner = getCurrentPartner( state );
+		isAgency = isAgencyUser( partner );
+	}
+	return isAgency;
+};
```
Comment (rcanepa) [inline client/lib/partner/utils.js]: I suggest moving this to `client/state/selectors/`, which is the place where we put selectors.

Here are additional details: https://github.com/Automattic/wp-calypso/blob/trunk/client/state/selectors/README.md

Comment (yashwin) [reply client/lib/partner/utils.js]: Done ✅

**Comment by yashwin** [comment]
> This works much better now. However, I noticed that when I go from a single site section (e.g., Scan) to Switch Site > All My Sites, I see a quick flash of a header that says `You are not authorized to see this page`, which makes me think there is a race condition within the check that decides what to show to the user.

Good catch, fixed now 👍

File: client/components/site-selector/index.jsx [hunk 9]
```diff
@@ -472,6 +474,10 @@ const navigateToSite =
 			const site = getSite( state, siteId );
 			debug( 'getPathnameForSite', siteId, site );
 
+			if ( isJetpackCloud() && ! site && showAgencyDashboard( state ) ) {
+				return page.redirect( '/dashboard' );
+			}
+
```
Comment (rcanepa) [inline client/components/site-selector/index.jsx]: This component is generic enough to be used across different apps (e.g., WordPress.com and Jetpack cloud), so it is not optimal to handle this particular case for Jetpack cloud.

Wouldn't it be possible to move this logic up the call stack? I suspect that we might be able to find a better place in some component that is specific to Jetpack cloud.

Comment (yashwin) [reply client/components/site-selector/index.jsx]: I don't think we can find a better place than this since this redirection happens on click of Switch Site and we have the logic currently in the navigate function which gets called on click. This function also handles redirection to pages like `/backup`, `/scan` etc. Basically, we have to redirect the user to `/dashboard` before it gets redirected to other pages.

Comment (rcanepa) [reply client/components/site-selector/index.jsx]: Understood. I checked that these changes are not affecting WordPress.com's behavior in any way.

**Review by rcanepa — APPROVED** [review]
LGTM now.

Thanks for working on this!

**Comment by yashwin** [comment]
> LGTM now.
> 
> Thanks for working on this!

Thanks for reviewing 🙂

File: client/components/site-selector/index.jsx [hunk 10]
```diff
@@ -506,9 +511,9 @@ const navigateToSite =
 			}
 		}
 
-		function getSiteBasePath( site ) {
+		function getSiteBasePath( siteObj ) {
```
Comment (rcanepa) [inline client/components/site-selector/index.jsx]: What is the reason behind this change? I think `site` is a common name for the object that contains information about a site. When we refer to a specific site's piece of data, we normally mention that in the variable's name, for instance, `siteSlug` or `siteId`.

**Review by rcanepa — APPROVED** [review]
Besides the comment I left, everything works as expected and looks correct.

File: client/components/site-selector/index.jsx (hunk 10 above)
Comment (yashwin) [reply client/components/site-selector/index.jsx]: This change was made because we move the site to the parent scope. I realised this change is not required and we don't have to pass site since it is already declared in the parent scope - here is the change - https://github.com/Automattic/wp-calypso/pull/62857/commits/8d90b8e56e63ec5efab1cdc79b7c77ff83439b6d

**Review by rcanepa — APPROVED** [review]
I just gave this PR a new quick manual test and everything still works as expected.
